File: src/stack.go
package stack

import (
	"errors"
)

// Stack представляет стек строк
type Stack struct {
	data     []string
	capacity int
	head     int
}

// New создает новый стек с указанной емкостью
func New(capacity int) *Stack {
	return &Stack{
		data:     make([]string, capacity),
		capacity: capacity,
		head:     -1,
	}
}

// NewDefault создает новый стек с емкостью по умолчанию (30)
func NewDefault() *Stack {
	return New(30)
}

// Push добавляет элемент на вершину стека
func (s *Stack) Push(value string) error {
	if s.head == s.capacity-1 {
		return errors.New("stack is full")
	}
	
	s.head++
	s.data[s.head] = value
	
	return nil
}

// Pop удаляет и возвращает элемент с вершины стека
func (s *Stack) Pop() (string, error) {
	if s.IsEmpty() {
		return "", errors.New("stack is empty")
	}
	
	value := s.data[s.head]
	s.head--
	
	return value, nil
}

// Peek возвращает элемент с вершины стека без удаления
func (s *Stack) Peek() (string, error) {
	if s.IsEmpty() {
		return "", errors.New("stack is empty")
	}
	
	return s.data[s.head], nil
}

// IsEmpty проверяет, пуст ли стек
func (s *Stack) IsEmpty() bool {
	return s.head == -1
}

// Size возвращает текущее количество элементов в стеке
func (s *Stack) Size() int {
	return s.head + 1
}

// Capacity возвращает емкость стека
func (s *Stack) Capacity() int {
	return s.capacity
}

// Clear очищает стек
func (s *Stack) Clear() {
	s.head = -1
}

File: src/stack.cpp
#include "../include/stack.h"

Stack::Stack(size_t size) : capacity(size), head(-1) {
    if(capacity == 0) capacity = 30;
    data = new string[capacity];
}

Stack::Stack() : capacity(30), head(-1) {
    data = new string[capacity];
}

Stack::~Stack() {
    delete[] data;
}

void Stack::push(string value) {
    if (head == capacity - 1) {
        // Автоматическое расширение
        size_t newCapacity = capacity * 2;
        string* newData = new string[newCapacity];
        
        for(size_t i = 0; i <= head; i++) {
            newData[i] = data[i];
        }
        
        delete[] data;
        data = newData;
        capacity = newCapacity;
    }
    data[++head] = value;
}

string Stack::pop() {
    if (head == -1) {
        throw underflow_error("Stack is empty");
    }
    return data[head--];
}

string Stack::peek() {
    if (head == -1) {
        throw underflow_error("Stack is empty");
    } 
    return data[head];
}

bool Stack::isEmpty() {
    return head == -1;
}

bool Stack::isFull() {
    return head == capacity - 1;
}

size_t Stack::size() {
    return head + 1;
}

// НОВЫЙ МЕТОД: очистка стека
void Stack::clear() {
    head = -1;
}

// НОВЫЙ МЕТОД: получение емкости
size_t Stack::getCapacity() {
    return capacity;
}
